### ISSUE TITLE
oauth2.0_token on Facebook returns error

### ISSUE THREAD

**Issue by tokumotion** [hadley/httr#324]
Hi, I've been trying to get my Facebook app token from `httr` using this code but got this message:

    > app <- oauth_app('facebook', mktg_api_ID, mktg_api_secret)
    > oauth2.0_token(oauth_endpoints("facebook"), app,
    + scope = '',
    + type  = "application/x-www-form-urlencoded")
    <Token>
    <oauth_endpoint>
     authorize: https://www.facebook.com/dialog/oauth
     access:    https://graph.facebook.com/oauth/access_token
    <oauth_app> facebook
      key:    444682622372984
      secret: <hidden>
    <credentials> {"error":{"message":"Error validating verification code. Please make sure your redirect_uri is identical to the one you used in the OAuth dialog request","type":"OAuthException","code":100,"fbtrace_id":"G+hjBKwmgHS"}}

Is this a package error or an app error?

**Comment by hadley** [comment]
It's most likely that you didn't set the redirect url to `oauth_callback()`

**Comment by tokumotion** [comment]
@Hadley I'm sorry, but how do I do that? Which is the redirect URL in this case?

**Comment by hadley** [comment]
The redirect url is what you get when you run `oauth_callback()`

**Comment by tokumotion** [comment]
@hadley Set the redirect URL to the value `oauth_callback()` returns and got the same error. It opens the browser in `http://localhost:1410/` with a success message, but R console returns error.

**Comment by msjgriffiths** [comment]
I also experienced this error. 

The reason is that `oauth_callback()` return e.g. `http://localhost:1410`; but Facebook does not allow a URL without a trailing slash. That is, the URL *should* be `http://localhost:1410/`. If you attempt to enter `http://localhost:1410` in the Facebook app settings, it **adds** the trailing slash.

I resolved this by monkey patching the function `oauth_callback`.

```r
library(httr)
library(R.utils)
custom_oauth_callback <- function(){
    paste0("http://", Sys.getenv("HTTR_SERVER", "localhost"), 
                ":", Sys.getenv("HTTR_SERVER_PORT", "1410"), "/")
}
reassignInPackage("oauth_callback", "httr", custom_oauth_callback)
```

From there, it works fine.
